### ISSUE TITLE
bug or not: qwen3 grpo training on search-r1 like tool calling, after 60 step I get enormous $ sign in the response during inference

### ISSUE THREAD

**Issue by lebronjamesking** [verl-project/verl#2819]
Hi, I am using qwen3-4B and https://github.com/volcengine/verl/blob/main/examples/sglang_multiturn/search_r1_like/run_qwen2.5-3b_instruct_search_multiturn.sh as the grpo training script. I get better results at step25 than before GRPO training (10k training sample and batch_size=128). However, after step 60, I get bad response as after tool_call with successful tool_response, then infinite $ $ $ $ is generated. Please can anyone tell me why?

Here is the training tensorboard log

<img width="1280" height="579" alt="Image" src="https://github.com/user-attachments/assets/dc8c4459-d50d-4fde-beab-a8ae3991ec6e" />

and here is infinite $ sign during inference

"Prompt": "<|im_start|>system\n\n            You are a helpful and harmless assistant. Task: Answer the given question by following these steps:\n\n1. Always begin your reasoning process by analyzing the question within <think> and </think> tags.\n\nYou are provided with function signatures within <tools></tools> XML tags:\nHere is function signature for code_interpreter:\n<tools>\n{\"type\": \"function\", \"function\": {\"name\": \"code_interpreter\", \"description\": \"A tool for executing code.\", \"parameters\": {\"type\": \"object\", \"properties\": {\"code\": {\"type\": \"string\", \"description\": \"The code to execute.\", \"enum\": null}}, \"required\": [\"code\"]}, \"strict\": false}}\n</tools>\nand here is function signature for search\n<tools>\n{\"type\": \"function\", \"function\": {\"name\": \"search\", \"description\": \"Searches the web for relevant information based on the given query.\", \"parameters\": {\"type\": \"array\", \"properties\": {\"query_list\": {\"type\": \"string\", \"description\": \"Searches the web for relevant information based on the given query.\", \"enum\": null}}, \"required\": \"query_list\"}, \"strict\": false}}\n</tools>\n\nFor each function call, return a json object with function name and arguments within <tool_call></tool_call> XML tags:\n<tool_call>\n{\"name\": <function-name>, \"arguments\": <args-json-object>}\n</tool_call>\n\n2. If additional information is needed:\n   - Perform a search by wrapping your query in <tool_call>query</tool_call>\n   - You may search multiple times as needed\n   - Results will appear between <tool_response> and </tool_response>\n\n3. For calculations or data processing:\n   - Write complete Python scripts including all necessary imports\n   - Wrap code in:\n     <tool_call>\n     ```python\n     # Your complete code here, use print() to show your final output\n\n     ```\n     </tool_call>\n   - Execution results will appear between <tool_response> and </tool_response>\n\n4. When ready to answer:\n   - If no further information is needed, provide just the answer between <answer> and </answer>\n   - Example: <answer>Beijing</answer>\n\nKey Principles:\n- Think step-by-step before taking any action\n- Use tools flexibly (search + code) as needed\n- Keep answers concise when possible\n- Maintain clear XML-style tagging throughout\n\nQuestion:\n\n        <|im_end|>\n<|im_start|>user\nWho is the father of the scientist at MIT that won the Queen Elizabeth Prize for Engineering in 2013?<|im_end|>\n<|im_start|>assistant\n",
        "Full_output": "<think>\nOkay, let's see. The question is asking for the father of the scientist at MIT who won the Queen Elizabeth Prize for Engineering in 2013. First, I need to figure out who that scientist was. The Queen Elizabeth Prize for Engineering is awarded every year, so I should check the list of recipients from 2013. \n\nI remember that the prize is given to individuals or teams who have made significant contributions to engineering. Maybe I can search for the 2013 winner. Let me try using the search function. Wait, the user mentioned that if additional information is needed, I should perform a search. So, I'll start by searching for the Queen Elizabeth Prize for Engineering 2013 recipient.\n\nOnce I find out who the scientist was, then I need to find their father. But maybe the scientist's father is a well-known figure, or perhaps it's someone less known. Alternatively, maybe the father is also an engineer or has some notable achievements. However, since the question is about the father, I might need to look up the family background of the 2013 winner.\n\nAlternatively, maybe the scientist is a public figure whose father is known. For example, if the scientist is a prominent figure like a Nobel laureate or a famous inventor, their father might be a notable person. But I need to confirm this. Let me first find out who the 2013 winner was.\n</think>\n\n<tool_call>\n{\"name\": \"search\", \"arguments\": {\"query_list\": \"Queen Elizabeth Prize for Engineering 2013 recipient MIT\"}}\n</tool_call><tool_response>\n\"Queen Elizabeth Prize for Engineering\"\nQueen Elizabeth Prize for Engineering The Queen Elizabeth Prize for Engineering, also known as the QEPrize, is a global engineering prize that rewards and celebrates the engineers responsible for a ground-breaking innovation in engineering that has been of global benefit to humanity. The £1 million Prize is awarded biennially in the name of Queen Elizabeth II. The prize was officially launched in November 2011 by a cross-party line up of the UK Prime Minister David Cameron, Deputy Prime Minister Nick Clegg and Leader of the Opposition Ed Miliband. The inaugural prize was presented to the creators of the Internet, the\n\n\"Queen Elizabeth Prize for Engineering\"\nallows users to export the design as a 3D print-ready .OBJ file. Queen Elizabeth Prize for Engineering The Queen Elizabeth Prize for Engineering, also known as the QEPrize, is a global engineering prize that rewards and celebrates the engineers responsible for a ground-breaking innovation in engineering that has been of global benefit to humanity. The £1 million Prize is awarded biennially in the name of Queen Elizabeth II. The prize was officially launched in November 2011 by a cross-party line up of the UK Prime Minister David Cameron, Deputy Prime Minister Nick Clegg and Leader of the Opposition Ed Miliband.\n\n\"Queen Elizabeth Prize for Engineering\"\n2013 prize were: On Tuesday 25 June 2013 the winners received their award from Her Majesty Queen Elizabeth II in front of an audience that included the three leaders of the UK’s main political parties, QEPrize judges and a number of young engineers at Buckingham Palace. On the same day, the Lord Mayor hosted a lunch at the Guildhall in honour of the winners attended by Princess Eugenie and Princess Beatrice. Key The QEPrize Global Engineering Ambassadors Network is an international network of young engineers from both business and academic institutions. QEPrize Ambassadors act as evangelists for engineering, engaging with\n\n\"Queen Elizabeth Prize for Engineering\"\nannouncement was made by Lord Browne of Madingley in the presence of HRH The Duke of York at the Royal Academy of Engineering on 3 February 2015. Dr Langer, who made a speech at the announcement, said he was “proud and privileged for winning the biggest engineering prize in the world”. The inaugural prize was awarded to a team of five engineers who created the Internet and the World Wide Web. The announcement was made by Lord Browne of Madingley in the presence of Princess Anne at the Royal Academy of Engineering on 18 March 2013. The winners of the\n\n\"Queen Elizabeth Prize for Engineering\"\nbrowser and the World Wide Web in 2013. The 2015 prize was awarded to Robert Langer, for his work in controlled-release large molecule drug delivery. The third prize was awarded on 1 February 2017 to Eric Fossum, George Smith, Nobukazu Teranishi and Michael Tompsett for research on digital imaging sensors. The objective of the QEPrize is to raise the public profile of engineering, and to inspire young people and engage them in engineering. It is run by a charitable company, limited by guarantee, called the Queen Elizabeth Prize for Engineering Foundation. The Foundation is chaired by Lord Browne of Madingley,\n\n\"Queen Elizabeth Prize for Engineering\"\ntree-like trophy design symbolises the growth of engineering and represents the way in which all areas of engineering are interlinked. After winning the competition Jennifer Leggett was invited to spend the day with designer Thomas Heatherwick before the design was finalised. The 2013 trophy was constructed out of a 3D printed resin that has been dipped in Nickel alloy. The tree-like trophy has some polished and some brushed modules, while the base is sintered aluminium. Entries to the 2013 competition were submitted through an online design website. For the first time since the launch of Create the Trophy, the 2017\n\n\"Queen Elizabeth Prize for Engineering\"\nannounced that the 2017 prize would be awarded to the four engineers responsible for the creation of digital imaging sensors. The announcement was made by Lord Browne of Madingley at the Royal Academy of Engineering, in the presence of HRH The Princess Royal. The winners of the 2017 prize were: Together, their innovations allowed for advancements in medical treatments, science, personal communications and entertainment. The winners will receive their award in a ceremony at Buckingham Palace later in 2017. The 2015 prize was awarded to Robert Langer of United States, for his work in controlled-release large molecule drug delivery. The\n\n\"Queen Elizabeth Prize for Engineering\"\nare announced every two years by the Chairman of the QEPrize Foundation. In the first three prize cycles, this announcement was held at the Royal Academy of Engineering, and was attended by members of the Royal Family. The QEPrize awards ceremony takes place in the same year as the announcement, at Buckingham Palace. The QEPrize trophy, designed by the winner of the Create the Trophy competition, is presented to the QEPrize winner(s) by a member of the Royal Family. In the first three prize cycles, the trophy was presented by Her Majesty The Queen. On 1 February 2017, it was\n\n\"T.H. Tse\"\non Debugging and Repair. He is an associate editor of \"\"IEEE Transactions on Reliability\"\" and an editorial board member of \"\"Software Testing, Verification and Reliability\"\" and \"\"Software: Practice and Experience\"\". He also served on the search committee for the editor-in-chief of \"\"IEEE Transactions on Software Engineering\"\" in 2013. He is a fellow of the British Computer Society. He was selected for a State Science and Technology Award, recommended by the Ministry of Education of P.R. China. He was also selected for an Outstanding Disabled Person of the Year Award. He was decorated with an MBE by Queen Elizabeth II. T.H.\n\n\"Queen Elizabeth Prize for Engineering\"\naudience, and received entries from 32 countries worldwide. The winner of the 2015 Create the Trophy competition was Euan Fairholm, a mechanical engineering student at The University of Glasgow. His design, \"\"The Golden Crown\"\", was developed into a final form by BAE Systems and presented to Dr Robert Langer, the winner of the 2015 QEPrize. Entries to the 2015 competition were submitted through an online design website. The inaugural Create the Trophy competition was won by 17-year-old student, Jennifer Leggett. The national competition called for students to come up with a design capturing the essence of modern engineering. Jennifer Leggett's\n</tool_response>\n\n$ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $"
        "Output": null,
        "answer": [
            "Conway Berners-Lee"
        ]
    },

my training hyper parameters are

# run on 8xH20
# make sure your current working directory is the root of the project
gpu_ids=0,1,2,3
export CUDA_VISIBLE_DEVICES=$gpu_ids
count=$(echo "$gpu_ids" | tr ',' '\n' | wc -l)
# export CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7
set -x

ulimit -n 65535

PROJECT_DIR="$(pwd)"
CONFIG_PATH="$PROJECT_DIR/examples/sglang_multiturn/config"

TRAIN_DATA="/kefu-nas/fuxinyu/verl_fxy/verl/examples/data_preprocess/tool_star_rl_data/ready/train.parquet"
VAL_DATA="/kefu-nas/fuxinyu/verl_fxy/verl/examples/data_preprocess/tool_star_rl_data/ready/test.parquet"

TOOL_CONFIG="$CONFIG_PATH/tool_config/search_tool_config.yaml"

python3 -m verl.trainer.main_ppo \
    --config-path="$CONFIG_PATH" \
    --config-name='search_multiturn_grpo' \
    algorithm.adv_estimator=grpo \
    data.train_batch_size=128 \
    data.val_batch_size=16 \
    data.max_prompt_length=2048 \
    data.max_response_length=8192 \
    data.filter_overlong_prompts=True \
    data.truncation='error' \
    data.return_raw_chat=True \
    actor_rollout_ref.model.path=/Qwen3-4B \
    actor_rollout_ref.actor.optim.lr=1e-6 \
    actor_rollout_ref.actor.optim.lr_warmup_steps_ratio=0.285 \
    actor_rollout_ref.model.use_remove_padding=True \
    actor_rollout_ref.actor.ppo_mini_batch_size=1 \
    actor_rollout_ref.actor.ppo_micro_batch_size_per_gpu=1 \
    actor_rollout_ref.actor.use_kl_loss=True \
    actor_rollout_ref.actor.kl_loss_coef=0.001 \
    actor_rollout_ref.actor.kl_loss_type=low_var_kl \
    actor_rollout_ref.actor.entropy_coeff=0 \
    actor_rollout_ref.model.enable_gradient_checkpointing=True \
    actor_rollout_ref.actor.fsdp_config.param_offload=False \
    actor_rollout_ref.actor.fsdp_config.optimizer_offload=False \
    actor_rollout_ref.rollout.max_model_len=16382 \
    actor_rollout_ref.rollout.log_prob_micro_batch_size_per_gpu=1 \
    actor_rollout_ref.rollout.tensor_model_parallel_size=1 \
    actor_rollout_ref.rollout.name=sglang \
    actor_rollout_ref.rollout.gpu_memory_utilization=0.6 \
    actor_rollout_ref.rollout.n=8 \
    actor_rollout_ref.rollout.multi_turn.max_assistant_turns=5 \
    actor_rollout_ref.ref.log_prob_micro_batch_size_per_gpu=1 \
    actor_rollout_ref.ref.fsdp_config.param_offload=True \
    algorithm.use_kl_in_reward=False \
    trainer.critic_warmup=0 \
    trainer.val_before_train=False \
    trainer.logger='["console","tensorboard"]' \
    trainer.project_name='search_r1_like_async_rl' \
    trainer.experiment_name='qwen3-4b-multitools' \
    trainer.n_gpus_per_node=$count \
    trainer.nnodes=1 \
    trainer.save_freq=5 \
    trainer.test_freq=50 \
    data.train_files="$TRAIN_DATA" \
    data.val_files="$VAL_DATA"  \
    actor_rollout_ref.rollout.multi_turn.tool_config_path="$TOOL_CONFIG" \
    trainer.total_epochs=2 $@

**Comment by lebronjamesking** [comment]
I am not sure if I should add stop tokens explicitely at sglang_rollout.py. For instance, somewhere after tool_call being executed. The grpo training data I use is suggesting max_response_length=4k and max_prompt_length=2k, so in total its 6k. However, in my setting I use something 2k+8k=10k. Do you think it's the reason why I get overfit after step60. 

Indeed, I think after step 30, the performance on test set bamboogle is already downgrading.

**Comment by Kirrito-k423** [comment]
定时清理检查结论：按社区 issue 定时处理规则关闭此 issue。

证据：最近一次非当前运行账号的公开回复或引用是 2025-07-30T14:39:33Z（actor: lebronjamesking, kind: comment），早于本轮 6 个月阈值 2026-01-13T01:10:17.218155+00:00。
检查来源：https://github.com/verl-project/verl/issues/2819，comments API https://api.github.com/repos/verl-project/verl/issues/2819/comments，timeline API https://api.github.com/repos/verl-project/verl/issues/2819/timeline。

如果作者或社区成员仍能提供新的复现、证据或需求背景，请在这里重新评论请求维护者重开，或创建新的 issue 并链接到本 issue。

Scheduled cleanup conclusion: this issue is being closed under the community issue processing policy.

Evidence: The latest public reply or reference from an account other than the current automation account was 2025-07-30T14:39:33Z (actor: lebronjamesking, kind: comment), which is older than the six-month threshold for this run (2026-01-13T01:10:17.218155+00:00).
Sources checked: https://github.com/verl-project/verl/issues/2819, comments API https://api.github.com/repos/verl-project/verl/issues/2819/comments, timeline API https://api.github.com/repos/verl-project/verl/issues/2819/timeline.

If the author or community can still provide a fresh reproduction, evidence, or requirement context, please comment here to ask maintainers to reopen it, or open a new issue and link back to this one.

This response was generated by a scheduled Codex GPT5.5 issue processing task.
